wolfcrypt/src/random.c

```diff
@@ -800,6 +809,779 @@ int wc_RNG_TestSeed(const byte* seed, word32 seedSz)
+        /* Hash the first 32 64-bit words of state. */
+        wc_Sha3_256_Update(&entropyHash, (byte*)entropy_state,
+            sizeof(*entropy_state) * ENTROPY_NUM_64BIT_WORDS);
+        /* Get pseudo-random indices. */
+        wc_Sha3_256_Final(&entropyHash, d);
+
+        for (i = 0; i < ENTROPY_NUM_64BIT_WORDS; i++) {
+            /* Choose a 64-bit word from a pseudo-random block.*/
+            int idx = ((int)d[i] << ENTROPY_BLOCK_SZ) +
+                      (j << ENTROPY_OFFSET_SHIFTING);
+            /* Update a pseudo-random 64-bit word with a pseudo-random value. */
+            entropy_state[idx] += Entropy_TimeHiRes();
+            /* Ensure part of state that is hashed is updated. */
+            entropy_state[i] += entropy_state[idx];
+        }
+    }
+}
+
+
+/* Last time entropy sample was gathered. */
+static word64 entropy_last_time = 0;
+
+/* Get a sample of noise.
+ *
+ * Value is time taken to use memory.
+ *
+ * Called to test raw entropy.
+ *
+ * @return  64-bit value that is the noise.
+ */
+static word64 Entropy_GetSample()
+{
+    word64 now;
+    word64 ret;
+
+    /* Use memory such that it will take an unpredicitable amount of time. */
+    Entropy_MemUse();
+
+    /* Get the time now to subtract from previous end time. */
+    now = Entropy_TimeHiRes();
+    /* Calculate time diff since last sampling. */
+    ret = now - entropy_last_time;
+    /* Store last time. */
+    entropy_last_time = now;
+
+    return ret;
+}
+
+/* Get as many samples of noise as required.
+ *
+ * One sample is one byte.
+ *
+ * @param [out] noise    Buffer to hold samples.
+ * @param [in]  samples  Number of one byte samples to get.
+ */
+static void Entropy_GetNoise(unsigned char* noise, int samples)
+{
+    int i;
+
+    /* Do it once to get things going. */
+    Entropy_MemUse();
+
+    /* Get as many samples as required. */
+    for (i = 0; i < samples; i++) {
+       noise[i] = (byte)Entropy_GetSample();
+    }
+}
+
+/* Generate raw entropy for performing assessment.
+ *
+ * @param [out] raw  Buffer to hold raw entropy data.
+ * @param [in]  cnt  Number of bytes of raw entropy to get.
+ * @return  0 on success.
+ * @return  Negative when creating a thread fails - when no high resolution
+ * clock available.
+ */
+int wc_Entropy_GetRawEntropy(unsigned char* raw, int cnt)
+{
+    int ret = 0;
+
+#ifdef ENTROPY_MEMUSE_THREADED
+    /* Start the counter thread as a proxy for time counter. */
+    ret = Entropy_StartThread();
+    if (ret == 0)
+#endif
+    {
+        Entropy_GetNoise(raw, cnt);
+    }
+#ifdef ENTROPY_MEMUSE_THREADED
+    /* Stop the counter thread to avoid thrashing the system. */
+    Entropy_StopThread();
+#endif
+
+    return ret;
+}
+
+#if ENTROPY_MIN == 1
+/* SP800-90b 4.4.1 - Repetition Test
+ * C = 1 + upper(-log2(alpha) / H)
+ * When alpha = 2^-30 and H = 1,
+ * C = 1 + upper(30 / 1) = 31
+ */
+#define REP_CUTOFF           31
+#else
+#error "Minimim entropy not defined to a recongnized value."
```

Comment:
Spelling: `#error "Minimum entropy not defined to a recognized value."`

Reply:
Fixed